server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -367,4 +369,95 @@ private static String indexMetadataFileName(IndexMetadata indexMetadata) {
         return String.join(DELIMITER, "metadata", String.valueOf(indexMetadata.getVersion()), String.valueOf(System.currentTimeMillis()));
     }
 
+    /**
+     * Fetch latest index metadata from remote cluster state
+     * @param clusterUUID uuid of cluster state to refer to in remote
+     * @param clusterName name of the cluster
+     * @return {@code Map<String, IndexMetadata>} latest IndexUUID to IndexMetadata map
+     */
+    public Map<String, IndexMetadata> getLatestIndexMetadata(String clusterUUID, String clusterName) throws IOException {
+        Map<String, IndexMetadata> remoteIndexMetadata = new HashMap<>();
+        ClusterMetadataManifest ClusterMetadataManifest = getLatestClusterMetadataManifest(clusterUUID, clusterName);
+        for (UploadedIndexMetadata uploadedIndexMetadata : ClusterMetadataManifest.getIndices()) {
+            IndexMetadata indexMetadata = getIndexMetadata(clusterUUID, clusterName, uploadedIndexMetadata);
+            remoteIndexMetadata.put(uploadedIndexMetadata.getIndexUUID(), indexMetadata);
+        }
+        return remoteIndexMetadata;
+    }
+
+    private IndexMetadata getIndexMetadata(String clusterUUID, String clusterName, UploadedIndexMetadata uploadedIndexMetadata)
+        throws IOException {
+        try {
+            return INDEX_METADATA_FORMAT.read(
+                indexMetadataContainer(clusterName, clusterUUID, uploadedIndexMetadata.getIndexUUID()),
+                uploadedIndexMetadata.getUploadedFilename(),
+                blobStoreRepository.getNamedXContentRegistry()
+            );
+        } catch (IOException e) {
+            String errorMsg = String.format(
+                Locale.ROOT,
+                "Error while downloading IndexMetadata - %s",
+                uploadedIndexMetadata.getUploadedFilename()
+            );
+            logger.error(errorMsg, e);
+            throw new IllegalStateException(errorMsg);
```

Comment:
set the innerCause as well. Did you check if this is resulting in error being logged twice? you can simply throw.

Reply:
Applies at all the places where you are logging and rethrowing error

Reply:
just added logger.error to ensure the message is not missed if the caller is not catching and then logging properly. this will ensure we will always log the error. let me know if you think its unnecessary, will revert

Reply:
I think this will result in logging twice, can you cross check and simply throw from all places.

Reply:
sure